### DISCUSSION TITLE
Anonymous Type - Return method

### DISCUSSION THREAD

**Discussion by boukenka** [dotnet/csharplang#4343] [Q&A]
Hello,

Recently, I wanted to improve the field selection in Entity Framework Core when selecting data.
I wrote the following code:

```C#
public async Task<(long)> GetProviderIdAsync(long id)
{
  return await _zzContext.Test
               .Where(x => x.Id == id)
               .Select(x => new { x.ProviderId })
               .FirstOrDefaultAsync();
}
```

Of course, it doesn't work. 😄 
I receive the following error message :

```
CS0029	Cannot implicitly convert type '<anonymous type: long ProviderId>' to '(long,?)'
```

Is it planned to implement in a future version of C#, an inference interpretation/logic to solve this issue ?
And not to manipulate it with a Dynamic type, an Object, another Anonymous type or Reflection?

For now, my solution is the following with a local variable:

```C#
public async Task<long> GetProviderIdAsync(long id)
{
    var result = await _zzContext.Test
                       .Where(x => x.Id == id)
                       .Select(x => new { x.ProviderId})
                       .FirstOrDefaultAsync();
    return result.ProviderId;
}
```

Thanks for any answer. 😄

**Comment by boukenka** [comment]
Why is it moved to discussion ?

**Reply by spydacarnage** [reply]
Because that is where the rules say questions of this type live.  Issues are for active proposals only.

**Reply by boukenka** [reply]
I wanted to ask that in the future, an anonymous object could be returned via a method in C# 10. So it needs an evolution of the C# langage.

**Reply by spydacarnage** [reply]
_Needs_ is a strong word.  You can return `object` already...

**Reply by boukenka** [reply]
Sorry. I am not a native speaker in English. I do not want to be _rude_ or _impolite_. 😄 Yes ```object``` can be used. But from my point, this is not an _elegant_ solution.

**Reply by spydacarnage** [reply]
How else would you want to specify an anonymous object as a return type?  The point of an anonymous type/object is that it has no definition (apart from when it is explicitly created) so you can't return a "typed" anonymous type.

`object` is the only solution.

**Reply by boukenka** [reply]
Ok. Thanks for the answer.

**Comment by spydacarnage** [comment]
What are you actually trying to return in the Task?  A `long` or a `(long)` ?
Regardless of that answer, neither is going to be able to be converted to from an anonymous object - why not just `.Select(x => x.ProviderId)` and return the long on its own?

**Reply by boukenka** [reply]
I was actually trying to return```(long)```. For sure, your proposition is better. But it will not work when you have multiple fields to return. Right ?

**Reply by spydacarnage** [reply]
If you're returning a tuple, create a tuple, so `.Select(x => (x.ProviderId))`, which can then return more than one item if you desire, like `.Select(x => (x.ProviderId, x.SomethingElse))`

**Reply by boukenka** [reply]
It will not work if the method returns ```Task<long, long>``` if SomethingElse is long type. ```Task<(long, long)>``` as well.

**Reply by spydacarnage** [reply]
No, then you would need to return `Task<(long, long)>`, like you originally had

**Reply by spydacarnage** [reply]
Or, use `record`s

**Reply by boukenka** [reply]
I have tried ```Task<(long, long)>```. It does not work. Here is the modified example from my original code:

```C#
public async Task<(long, long)> GetByProviderIdAsync(long id)
{
    return await _zzContext.Test
        .Where(x => x.Id == id)
        .Select(x => (x.ProviderId, x.ProviderId))
        .FirstOrDefaultAsync();
}
```

**Reply by boukenka** [reply]
I get a ```CS8143``` error for the Select.

**Reply by boukenka** [reply]
So I have modified the code to remove the error. We are back to the original requirement.

```C#
public async Task<(long, string)> GetProviderIdAsync(long id)
{
    return await _zzContext.Test
        .Where(x => x.Id == id)
        .Select(x => new { x.ProviderId, x.Message })
        .FirstOrDefaultAsync();
}
```

Error is : ```CS0029	Cannot implicitly convert type '<anonymous type: long ProviderId, string Message>' to '(long, string)```

**Reply by boukenka** [reply]
@spydacarnage  Do you have any comment about it ? 😄

**Reply by boukenka** [reply]
I have found another solution:

```C#
public async Task<Tuple<long, string>> GetProviderIdAsync(long id)
{
    return await _zzContext.Test
        .Where(x => x.Id == id)
        .Select(x => new Tuple<long, string>(x.ProviderId, x.Message))
        .FirstOrDefaultAsync();
}
```

**Reply by boukenka** [reply]
Another solution like the one you suggested earlier :

```C#
public record Data(long ProviderId, string Message);

public async Task<Data> GetProviderIdAsync(long providerId)
{
    return await _zzContext.Test
        .Where(x => x.Id== id)
        .Select(x => new Data(x.ProviderId, x.Message))
        .FirstOrDefaultAsync();
}
```

**Reply by boukenka** [reply]
Thank you for the discussion. 😄

**Reply by Joe4evr** [reply]
> I get a CS8143 error for the Select.

This is because ValueTuples are not supported in expression lambdas. See also: #2545.

**Reply by boukenka** [reply]
Thank you @Joe4evr. Seems there is no real progress in #2545.

**Reply by yaakov-h** [reply]
The expressions library is feature-locked. It is unlikely that expressions will ever represent new language constructs.

**Reply by boukenka** [reply]
Thank for the information @yaakov-h 😄
